src/Workspaces/Core/Portable/Workspace/Solution/DocumentState.cs

```diff
@@ -457,6 +449,97 @@ public DocumentState UpdateAnalyzerConfigSet(ValueSource<AnalyzerConfigSet> newA
                 newTreeSource);
         }
 
+        private ValueSource<TreeAndVersion>? CreateNewTreeSourceForUpdatedAnalyzerConfigSet(ValueSource<AnalyzerConfigSet> newAnalyzerConfigSet)
+        {
+            if (!SupportsSyntaxTree)
+                return null;
+
+            var filePath = GetSyntaxTreeFilePath(this.Attributes);
+
+            // If we don't have an existing tree, there's no optimization we can do.  Just lazily
+            // parse the tree fully when it is requested.
+            if (_treeSource == null)
+                return CreateLazyFullyParsedTree(newAnalyzerConfigSet, filePath);
+
+            // In the case where the config file didn't change between the last version
+            // and this version (in any significant way) we want to reuse the previous
+            // document to prevent excess unnecessary work.
+
+            return new AsyncLazy<TreeAndVersion>(
+                c => FullyParseTreeIfChangedByUpdatedConfigAsync(_treeSource, newAnalyzerConfigSet, filePath, c),
+                c => FullyParseTreeIfChangedByUpdatedConfig(_treeSource, newAnalyzerConfigSet, filePath, c),
+                cacheResult: true);
+        }
+
+        private ValueSource<TreeAndVersion> CreateLazyFullyParsedTree(ValueSource<AnalyzerConfigSet> newAnalyzerConfigSetSource, string filePath)
+            => CreateLazyFullyParsedTree(
+                TextAndVersionSource, Id.ProjectId, filePath, _options!, newAnalyzerConfigSetSource, _languageServices);
+
+        /// <summary>
+        /// Asynchronous version of <see cref="FullyParseTreeIfChangedByUpdatedConfig"/>.
+        /// Keep the code for both in sync.
+        /// </summary>
+        private async Task<TreeAndVersion> FullyParseTreeIfChangedByUpdatedConfigAsync(
```

Comment:
Does this need to be static and explicitly be passed in the old analyzer config set source? Otherwise this is still rooting the old DocumentState through the implicit `this`, right?